CRAMv3.tex

```diff
@@ -1336,16 +1336,22 @@ \subsection{Read names (RN data series)}
 \subsection{\textbf{Mate record}}
 \label{subsec:mate}
 
-There are two ways in which mate information can be preserved in CRAM: number of records downstream (distance, within this slice) to the next fragment in the template and a special mate record if the next fragment is not in the current slice.
-In the latter case the record is labelled as ``detached'', see the CF data series.
+There are two ways in which mate information can be preserved in CRAM.
+If the next fragment is not in the same slice we store verbatim copies of the insert size, mate reference chromosome and positions, and mate flags (mapped status, orientation) for both records.
+In this case both records are labelled as ``detached'' in the CF data series using bit 2.
 
-For mates within the slice only the distance is captured, and only for the first record.  The mate has neither detached nor downstream flags set in the CF data series.
+If this and the next fragment are within the same slice, we can derive much of this information by comparing the two records.
+The upstream record has CF bit 4 (mate downstream) flag set and stores the number of records to skip (in the NF data series) between this record and the record for the next fragment on this template, with zero meaning the next fragment is also the next record.
+The downstream record has neither CF bits 2 (detached) or 4 (mate downstream) set nor does it use the NF data series (unless it also has an additional ``next fragment'' to refer to).
```

Comment:
What is the implication for how the downstream record is stored in this case (both in the same slice) ? The pseudo code seems to imply that one or the other of the records don't have the doesn't have redundant data (insert size, mate reference chromosome and positions, and mate flags (mapped status, orientation) stored ?

Reply:
I think that comment is confusing.

We never remove entire records, so if the SAM has 1 million records so will the CRAM.  However the CF bit flags indicate whether specific columns in those records are stored verbatim or derived.  Thus the number of items in each column may differ to the total record count, even for mandatory columns such as TLEN.

I don't see what the problem is with the pseudocode here.  Possibly it's incomplete, but I think it was derived from my implementation (which was likely in turn derived partly from the original spec and partly from the Java code).  For example:

C:
```
                // set mate reversed if needed
                if (s->crecs[cr->mate_line].flags & BAM_FREVERSE)
                    cr->flags |= BAM_FMREVERSE;
```

Java:
```
        if ((cramMateFlags & MATE_NEG_STRAND_FLAG) != 0) value |= BAM_FLAGS.MATE_STRAND_FLAG;
```

Pseudocode:
```
if next_frag.bam_flags AND 0x10 then
    this.bam_flags ← this.bam_flags OR 0x20                   // next segment reverse complemented
end if
```

I could see the logic being explicit in both ways.  Ie what it misses right now in implementations is the else clause.  Eg:

```
                if (s->crecs[cr->mate_line].flags & BAM_FREVERSE)
                    cr->flags |= BAM_FMREVERSE;
               else
                    cr->flags &= ~BAM_FMREVERSE;
```

The assumption though is that you're not going to encode a BF data series element with needless bits set only to unset them algorithmically later.  Our implementations don't do that, hence I didn't document it as such.

Note there's nothing that specifically states we MUST clear those flags either.  I see this tiny little test works differently between C and Java implementations:

```
a1	99	xx	1	1	10M	=	11	20	AAAAAAAAAA	**********
a1	147	xx	11	1	10M	=	1	-20	TTTTTTTTTT	**********
```

The C code just stores 99 and 147 as the BF field.  The Java code stores 67 and 147.  Ie it clears MREVERSE as it knows it'll be adding it back.  Cleverer. :-)